### ISSUE TITLE
Usecase: overriding constants

### ISSUE THREAD

**Issue by oprypin** [crystal-lang/crystal#3241]
Consider this example, where I would like subclasses to be able to override some behavior by setting a constant with a different value than in the base class.

``` crystal
class O
  def initialize(@x : Int32)
    puts "initialized(#{@x})"
  end
end

class A
  V = O.new(6)
  def test
    V
  end
end

class B < A
  V = O.new(7)
end

class C < B
end

pp A::V, B::V, C::V
pp A.new.test, B.new.test, C.new.test
```

```
initialized(6)
initialized(7)
A::V # => #<O:0x2369f50 @x=6>
B::V # => #<O:0x2369f40 @x=7>
C::V # => #<O:0x2369f40 @x=7>
A.new.test # => #<O:0x2369f50 @x=6>
B.new.test # => #<O:0x2369f50 @x=6>
C.new.test # => #<O:0x2369f50 @x=6>
```

Initialization and usage of this constant works as expected, but when using a constant inside a method, the behavior is surprising. I understand that the method is tied to the value of the constant at the moment of definition and has no reason to change its behavior in a subclass, but this behavior would be nice to have, especially because there is no good alternative as far as I can see. Bad alternatives:
- Using a method instead of a constant
  - Gets initialized every time
- Accompanying the constant with a method every time it's redefined
  - Too annoying to use
- Using a class variable instead
  - Does not really represent the fact that this is a constant; access is hindered. But, more importantly, it is initialized once per class which may not be wanted.
- Using `{% begin %}{{@type.id}}::V{% end %}` instead of `V`
  - ...
- Ruby's `self.class::V`
  - Does not work

**Comment by asterite** [comment]
I think `{% begin %}{{@type.id}}::V{% end %}` is fine, I wouldn't add syntax or another way to do something that can be currently done, and is not very commonly needed.

We even use it in `Int`: https://github.com/crystal-lang/crystal/blob/5887fd69c49a8767a71fad74e83fdbc979897058/src/int.cr#L141-L145

**Comment by BlaXpirit** [comment]
The problem is it's surprising behavior. Every other kind of unqualified access (methods, instance variables, class variables) always refers to the current class, and not some base class. If I really wanted the constant from one particular class, I would specify that class directly.

I do not suggest another syntax, but doing this by default with just `V` instead. I understand that it may make compilation more complicated, but it does not affect cases where constants are never overridden.

How can you be sure it is not very commonly needed? It's a widespread idiom in Python, and I find it even more important in Crystal, where the run-time alternatives to this pattern don't work.

**Comment by BlaXpirit** [comment]
And to back this up with an example:

```crystal
class Demo
  FONT = SF::Font.from_file("resources/font/Cantarell-Regular.otf")
  TITLE = "crystal-chipmunk"
  SIM_FPS = 60

class TheoJansen < Demo
  TITLE = "Theo Jansen Machine"
  SIM_FPS = 180
```

No, I don't think having `{% begin %}{{@type.id}}::V{% end %}` all over the place is acceptable, and I may want to add even more of these configurable values.

**Comment by asterite** [comment]
Could you show an equivalent python program?

**Comment by BlaXpirit** [comment]
```python
class O:
    def __init__(self, x):
        self.x = x
        print("initialized({})".format(x))

    def __repr__(self):
        return '<{}>'.format(self.x)

class A:
    V = O(6)
    def test(self):
        return self.V

class B(A):
    V = O(7)

class C(B):
    pass

print(A.V, B.V, C.V)
print(A().test(), B().test(), C().test())
```

```
initialized(6)
initialized(7)
<6> <7> <7>
<6> <7> <7>
```

**Comment by bcardiff** [comment]
Maybe `typeof(self)::V` is the crystal way to delay the binding of a constant. I've try it i am over mobile now.

**Comment by asterite** [comment]
> The problem is it's surprising behavior. Every other kind of unqualified access (methods, instance variables, class variables) always refers to the current class, and not some base class. If I really wanted the constant from one particular class, I would specify that class directly.

But in Python you must use `self.V`, not just `V`.

In your font-title-sim_fps example I would use instance vars and don't use subclasses. It's simpler to configure by passing these values to a constructor or method than to subclass an object.

We could consider adding `exp::Const` in the future, to make a dynamic constant lookup, similar to how method dispatch works, but for me it's super low priority right now because there's a way to do that, using `{{@type}}`.

`typeof(self)::V` is a good idea, but it doesn't parse.

**Comment by BlaXpirit** [comment]
> But in Python you must use `self.V`, not just `V`.

Well, Python doesn't have unqualified access at all. Just `V` means a global `V`. This is as close as it gets. And Crystal can't do a similar action in any reasonable way.

> In your font-title-sim_fps example I would use instance vars and don't use subclasses

You see just the top of my classes, but they also include overriden methods (surely the difference between two different demos is not just their FPS ^_^). How can you do that without subclassing?

**Comment by ozra** [comment]
The Python results shown above are what I would expect as "the natural resolution".
Of course the constant should be resolved in the context of the actual type, not where the methods where defined - imho.
I can't see how the opposite would be commonly wanted - for that case you'd simply use another constant in the subtypes, no?

**Comment by BlaXpirit** [comment]
Welp, throwing around `{% begin %}{{@type.id}}::V{% end %}` has been fun and all, but it's not so funny anymore when `@type.id == "BaseClass+"`

**Comment by RX14** [comment]
It seems there exists some cases where the macro solution doesn't cut it, so a new solution (macro or not) should probably be worked out.

@oprypin Can you provide some example code where `@type.id == "BaseClass+"`?

**Comment by z64** [comment]
@fgimian note that example only makes sense if you have `abstract class Animal`. (`Animal#make_noise` is broken if you try to instantiate it)

**Comment by fgimian** [comment]
Another approach is to keep it as a class method and access it using `class` on an instance:

```crystal
class Animal
  NOISE = "eeep"

  # Creates an instance method that expsoses the class constant
  macro create_make_noise_method
    def self.make_noise
      NOISE
    end
  end

  # Create the make_noise method for the Animal class
  create_make_noise_method

  # Create the make_noise method for all subclasses of Animal
  macro inherited
    create_make_noise_method
  end
end

class Dog < Animal
  NOISE = "bark"
end

animal = Animal.new
animal.class.make_noise # => eeep

dog = Dog.new
dog.class.make_noise # => bark
```

Or of course, you could define a class method and instance method that do the same thing if you like 😄 

Hope this helps
Fotis

**Comment by straight-shoota** [comment]
You can use a macro for this: `{% begin %}{{ @type }}::NOISE{% end %}`. Maybe there could also be a way to resolve `self.class.NOISE`.

**Comment by oprypin** [comment]
That was mentioned in the very first post, also later pointed out as not always working. Though it might always work now after #6024

**Comment by asterite** [comment]
The solution is given in the very first post:

> Accompanying the constant with a method every time it's redefined

I mean, I want my program to do what I want without writing any code at all. When are we going to fix that in Crystal?

Please use regular virtual methods and redefining methods in subclasses. Program against an interface, not a constant.

All the example aboves are super easy: "noise" is a method of Animal, and that should be redefined for each animal. You need to initialize it once? Just define a constant and return that in the method. The constant doesn't even need to be public, or named similarly in every subclass. In fact it's better not to rely on a constant: what if an Animal does noises depending on some state? Of course you can redefine a method then, but having "well, the easy route is to just redefine the constant, or if you want more power redefine the method" is more complex to explain and understand than just "redefine the method".

Could someone close this issue now? Redefining constants will **never** happen.

**Comment by ysbaddaden** [comment]
:+1: to asterite comment: all your problems are solved with a method. If the value really looks like a constant and is shared by all instances then use a class method (and maybe class variable).
